File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/processor/StatelessProcessor.java
package io.github.oleksivio.telegram.bot.core.controller.processor;

import io.github.oleksivio.telegram.bot.api.model.objects.Update;
import io.github.oleksivio.telegram.bot.core.controller.handler.Handler;
import org.springframework.stereotype.Controller;

import java.util.ArrayList;
import java.util.Collection;

@Controller
public class StatelessProcessor {

    private final Collection<Handler> handlers = new ArrayList<>();

    public void add(Handler handler) {
        handlers.add(handler);
    }

    void receive(Update update) {
        handlers.stream()
                .filter(handler -> handler.hasSubscription(update))
                .forEach(handler -> handler.run(update));

    }

}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/processor/StatelessProcessor.kt
package io.github.oleksivio.telegram.bot.core.controller.processor

import io.github.oleksivio.telegram.bot.api.model.objects.Update
import io.github.oleksivio.telegram.bot.core.controller.handler.Handler
import org.springframework.stereotype.Controller
import java.util.*

@Controller
class StatelessProcessor {

    private val handlers = ArrayList<Handler<*>>()

    fun add(handler: Handler<*>) {
        handlers.add(handler)
    }

    internal fun receive(update: Update) {
        handlers.filter { handler -> handler.hasSubscription(update) }
                .forEach { handler -> handler.run(update) }

    }

}
